aries_vcx/src/protocols/issuance_v2/issuer/mod.rs

```diff
@@ -0,0 +1,571 @@
+///         offer_preview,
+///         None,
+///     )
+///     .await
+///     .unwrap();
+/// ```
+///
+/// For more information about formats, see [IssuerCredentialIssuanceFormat] documentation.
+pub struct IssuerV2<S> {
+    state: S,
+    thread_id: String,
+}
+
+impl<S> IssuerV2<S> {
+    pub fn from_parts(thread_id: String, state: S) -> Self {
+        Self { state, thread_id }
+    }
+
+    pub fn into_parts(self) -> (String, S) {
+        (self.thread_id, self.state)
+    }
+
+    /// Get the thread ID that is being used for this protocol instance.
+    pub fn get_thread_id(&self) -> &str {
+        &self.thread_id
+    }
+
+    pub fn get_state(&self) -> &S {
+        &self.state
+    }
+}
+
+impl<T: IssuerCredentialIssuanceFormat> IssuerV2<ProposalReceived<T>> {
+    /// Initialize a new [IssuerV2] by receiving an incoming [ProposeCredentialV2] message from a
+    /// holder.
+    ///
+    /// The [IssuerCredentialIssuanceFormat] used during initialization should be suitable
+    /// for the attachments within the [ProposeCredentialV2] message, or else the [IssuerV2] will
+    /// not be able to transition forward without failure.
+    ///
+    /// This API should only be used for standalone proposals that aren't apart of an existing
+    /// protocol thread. Proposals in response to an ongoing thread should be handled via
+    /// [HolderV2::receive_proposal].
+    pub fn from_proposal(proposal: ProposeCredentialV2) -> Self {
+        IssuerV2 {
+            thread_id: get_thread_id_or_message_id!(proposal),
+            state: ProposalReceived {
+                proposal,
+                _marker: PhantomData,
+            },
+        }
+    }
+
+    /// Get the details and credential preview (if any) of the proposal that was received. The
+    /// returned [IssuerCredentialIssuanceFormat::ProposalDetails] data will contain data
+    /// specific to the format being used.
+    pub fn get_proposal_details(
+        &self,
+    ) -> VcxResult<(T::ProposalDetails, Option<&CredentialPreviewV2>)> {
+        let details = T::extract_proposal_details(&self.state.proposal)?;
+        let preview = self.state.proposal.content.credential_preview.as_ref();
+
+        Ok((details, preview))
+    }
+
+    /// Respond to a proposal by preparing a new offer. This API can be used repeatedly to negotiate
+    /// the offer with the holder until an agreement is reached.
+    ///
+    /// An offer is prepared in the format of [IssuerCredentialIssuanceFormat], using the provided
+    /// input data to create it. Additionally, a [CredentialPreviewV2] is attached to give further
+    /// details to the holder about the offer.
+    ///
+    /// In the event of failure, an error is returned which contains the reason for failure
+    /// and the state machine before any transitions. Consumers should decide whether the failure
+    /// is terminal, in which case they should prepare a problem report.
+    pub async fn prepare_offer(
```

Comment:
kk, latest commit converts to this approach, let me know what you think. Would be helpful to me if we could go over the reason _why_ again though..

I originally thought it was bcus we were switching to the mentality of "our state machines should only store the information required to keep transitioning", but that is not necessarily true, because i'm still storing other convenience items in the state machines: e.g. `CredentialReceived` state is also storing `stored_credential_metadata`, which is just a convenience for the state machine consumer so that they can do a `holder.get_stored_credential_metadata()` and see information about the credential they just stored.

If we were to go all the way with the "our state machines should only store the information required to keep transitioning" mentality, then `stored_credential_metadata` would be removed from `CredentialReceived` state. and instead it would be returned from the `HolderV2::receive_credential` API in a tuple like we do with messages.

Same story with the issuer's `CredentialPrepared` state, which stores `credential_metadata`, allowing them to do a `issuer.get_credential_creation_metadata()` and see information such as the cred-rev-id of the created credential. To remove `credential_metadata` from the state, then the `IssuerV2::prepare_credential` API would need to return this metadata AND the credential message:
```
pub async fn prepare_credential(self,...) -> <(IssuerV2<CredPrepared>, IssueCredentialV2Message, T::CreatedCredentialMetadata)>
```

So i guess it'd be good if we made a decision/statement about how far we want to lean into the "our state machines should only store the information required to keep transitioning" mentality